### PR TITLE
Core framework alpha: middlewares.

### PR DESCRIPTION
**Pull request by williamvinogradov**
Also, a little refactoring in  getKeys.ts utils.

### CODE REVIEW

File: packages/core/src/component/middlewares/changesMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import { getChangedKeys } from './getChangedKeys';
+
+type ModelChangesTuple<TModel extends ObjectType> = [newModel: TModel, hasChanges: boolean];
+
+export const changesMiddleware = <TModel extends ObjectType>(
+  prev: TModel,
+  next: TModel,
+  config: ModelConfigMap<TModel> = {},
+): ModelChangesTuple<TModel> => {
+  const result = { ...next };
```
Comment (MikeVitik) [inline packages/core/src/component/middlewares/changesMiddleware.ts:12]: Alternative variant:
```
const result = { ...next };
  const controlledKeys = getChangedKeys(prev, next).filter((key) => {
const configItem = config[key];

    return  (configItem && configItem.controlledMode)});
  controlledKeys.forEach((key) => {
      result[key] = prev[key];
    }
  });

  return [result, controlledKeys.lenght > 0];
```

**Review by MikeVitik — DISMISSED** [review]

File: packages/core/src/component/middlewares/changesMiddleware.ts (hunk 1 above)
Comment (williamvinogradov) [reply packages/core/src/component/middlewares/changesMiddleware.ts:12]: Discussed, changesMiddleware rewrited

**Review by MikeVitik — DISMISSED** [review]

File: packages/core/src/component/middlewares/__test__/callbacksMiddleware.test.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        controlledMode: false,
+        changeCallback: (value: number) => { result.push(value); },
+      },
+    };
+    getChangedKeysMock.mockReturnValue(['a', 'b', 'c']);
+
+    const funcArray = callbacksMiddleware(prev, next, config);
+    funcArray.forEach((func) => func());
+
+    expect(result).toEqual(expectedResult);
```
Comment (IlyaKhD) [inline packages/core/src/component/middlewares/__test__/callbacksMiddleware.test.ts:38]:
```suggestion
    const callback = fn();
    const config = {
      a: {
        changeCallback: callback,
      },
      c: {
        changeCallback: fn(),
      },
    };
    getChangedKeysMock.mockReturnValue(['a', 'b']);

    const actual = callbacksMiddleware({}, {}, config);

    expect(actual).toEqual([callback]);
```

File: packages/core/src/component/middlewares/changesMiddleware.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { ObjectType } from '../../utils';
+import { ModelConfigMap } from './types';
+import { getChangedKeys } from './getChangedKeys';
+
+type ModelChangesTuple<TModel extends ObjectType> = [newModel: TModel, hasChanges: boolean];
+
+export const changesMiddleware = <TModel extends ObjectType>(
```
Comment (IlyaKhD) [inline packages/core/src/component/middlewares/changesMiddleware.ts:7]:
```suggestion
export const revertControlledProps = <TModel extends ObjectType>(
```
or 

```suggestion
export const controlledModeMiddleware = <TModel extends ObjectType>(
```

File: packages/core/src/component/middlewares/types.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { ObjectType } from '../../utils';
+
+export interface ModelPropertyConfig<T> {
+  controlledMode: boolean;
```
Comment (IlyaKhD) [inline packages/core/src/component/middlewares/types.ts:4]: It seems that `controlledMode` is required only for the changesMiddleware. Can we move it away from this common type?

e.g.

```ts
    const changesMiddleware = createChangesMiddleware(controlledProps as Set<string>)
```

File: packages/core/src/utils/getKeys.ts [hunk 5]
```diff
@@ -1,3 +1,7 @@
 import { ObjectType } from './types';
 
-export const getKeys = <T>(object: ObjectType<T>) => Object.keys(object) as (keyof T)[];
+export const getKeys = <T extends ObjectType>(object: T) => {
+  const keys = Object.keys(object);
+  const symbols = Object.getOwnPropertySymbols(object);
+  return [...keys, ...symbols] as (keyof T)[];
+};
```
[the remark this answers is not in the document]
Comment (williamvinogradov) [reply packages/core/src/utils/getKeys.ts:7]: It looks cool and I agree with you.

File: packages/core/src/component/middlewares/changesMiddleware.ts (hunk 3 above)
Comment (williamvinogradov) [reply packages/core/src/component/middlewares/changesMiddleware.ts:7]: Okay, let's rename it with ```controlledModeMiddleware```

File: packages/core/src/component/middlewares/types.ts (hunk 4 above)
Comment (williamvinogradov) [reply packages/core/src/component/middlewares/types.ts:4]: Discussed in the live conversation.
It looks a little complicated to pick only the right values from the initial config Record.
So, my suggestion is to leave it as is for now.

File: packages/core/src/component/middlewares/__test__/callbacksMiddleware.test.ts (hunk 2 above)
Comment (williamvinogradov) [reply packages/core/src/component/middlewares/__test__/callbacksMiddleware.test.ts]: I simplified this test case ([link](https://github.com/williamvinogradov/multiplatform-spike/pull/11/files#diff-0777fea209254f30b49af1d5c597e8d8617a10f13cea4724a5f3c2ea924df560R18)).
Please, take a look

**Review by IlyaKhD — APPROVED** [review]

**Review by MikeVitik — APPROVED** [review]

**Pull request merged**